### PR TITLE
Speed up `brew tap` for no arguments

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This provides a >3x speedup for `brew tap` with no arguments (i.e., when listing taps). It also makes the completion significantly faster.

```console
$ hyperfine --warmup=3 --setup 'git checkout {branch}' --parameter-list branch master,brew-tap-speedup 'brew tap'
Benchmark 1: brew tap (branch = master)
  Time (mean ± σ):      1.405 s ±  0.080 s    [User: 0.561 s, System: 0.238 s]
  Range (min … max):    1.332 s …  1.549 s    10 runs

Benchmark 2: brew tap (branch = brew-tap-speedup)
  Time (mean ± σ):     404.1 ms ± 124.8 ms    [User: 107.9 ms, System: 200.7 ms]
  Range (min … max):   308.8 ms … 693.7 ms    10 runs

Summary
  brew tap (branch = brew-tap-speedup) ran
    3.48 ± 1.09 times faster than brew tap (branch = master)
```

### CODE REVIEW

**Review by Rylan12 — APPROVED** [review]
This is great!

File: Library/Homebrew/tap.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# HOMEBREW_LIBRARY is set by bin/brew
+# shellcheck disable=SC2154
+
+normalise_tap_name() {
+  local dir="$1"
+  local user
+  local repo
+
+  user="$(echo "${dir%%/*}" | tr '[:upper:]' '[:lower:]')"
+  repo="$(echo "${dir#*/}" | tr '[:upper:]' '[:lower:]')"
```
Comment (carlocab) [inline Library/Homebrew/tap.sh:11]:
```suggestion
  user="$(tr '[:upper:]' '[:lower:]' <<<"${dir%%/*}")"
  repo="$(tr '[:upper:]' '[:lower:]' <<<"${dir#*/}")"
```
Maybe? Tho probably doesn't matter much.

**Review by carlocab — APPROVED** [review]

File: Library/Homebrew/brew.sh [hunk 2]
```diff
@@ -172,6 +172,11 @@ case "$@" in
     source "${HOMEBREW_LIBRARY}/Homebrew/list.sh"
     homebrew-list "$@" && exit 0
     ;;
+  # falls back to cmd/tap.rb on a non-zero return
+  tap*)
```
Comment (carlocab) [inline Library/Homebrew/brew.sh:176]:
```suggestion
  tap)
```
I wonder if this makes more sense so we don't have to handle `tap*` in `homebrew-tap`

File: Library/Homebrew/tap.sh (hunk 1 above)
Comment (ZhongRuoyu) [reply Library/Homebrew/tap.sh]: Thanks for reminding me again of this feature 😂

File: Library/Homebrew/tap.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+normalise_tap_name() {
+  local dir="$1"
+  local user
+  local repo
+
+  user="$(tr '[:upper:]' '[:lower:]' <<<"${dir%%/*}")"
+  repo="$(tr '[:upper:]' '[:lower:]' <<<"${dir#*/}")"
+  repo="${repo#@(home|linux)brew-}"
+  echo "${user}/${repo}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.sh:13]:
```suggestion
  repository="$(tr '[:upper:]' '[:lower:]' <<<"${dir#*/}")"
  repository="${repository#@(home|linux)brew-}"
  echo "${user}/${repository}"
```

File: Library/Homebrew/tap.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Does the quickest output of brew tap possible for no arguments.
+# HOMEBREW_LIBRARY is set by bin/brew
+# shellcheck disable=SC2154
+
+normalise_tap_name() {
+  local dir="$1"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.sh:6]:
```suggestion
  local directory="$1"
```

File: Library/Homebrew/tap.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  local user
+  local repo
+
+  user="$(tr '[:upper:]' '[:lower:]' <<<"${dir%%/*}")"
+  repo="$(tr '[:upper:]' '[:lower:]' <<<"${dir#*/}")"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.sh:11]:
```suggestion
  user="$(tr '[:upper:]' '[:lower:]' <<<"${directory%%/*}")"
  repo="$(tr '[:upper:]' '[:lower:]' <<<"${directory#*/}")"
```

File: Library/Homebrew/brew.sh [hunk 6]
```diff
@@ -172,6 +172,11 @@ case "$@" in
     source "${HOMEBREW_LIBRARY}/Homebrew/list.sh"
     homebrew-list "$@" && exit 0
     ;;
+  # falls back to cmd/tap.rb on a non-zero return
+  tap)
+    source "${HOMEBREW_LIBRARY}/Homebrew/tap.sh"
+    homebrew-tap "$@" && exit 0
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:178]: How/when will this exit with a non-zero result? Not currently seeing any e.g. `return 1` in there?

**Review by MikeMcQuaid — COMMENTED** [review]
Great idea, thanks @ZhongRuoyu! A few more tweaks and then should be good to go.

File: Library/Homebrew/tap.sh [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+homebrew-tap() {
+  local taplib="${HOMEBREW_LIBRARY}/Taps"
+  (
+    shopt -s extglob
+
+    for dir in "${taplib}"/*/*
+    do
+      [[ -d "${dir}" ]] || continue
+      dir="${dir#"${taplib}"/}"
+      normalise_tap_name "${dir}"
```
Comment (ZhongRuoyu) [inline Library/Homebrew/tap.sh:25]:
```suggestion
    for directory in "${taplib}"/*/*
    do
      [[ -d "${directory}" ]] || continue
      normalise_tap_name "${directory#"${taplib}"/}"
```

File: Library/Homebrew/brew.sh (hunk 6 above)
Comment (ZhongRuoyu) [reply Library/Homebrew/brew.sh:177]: Sorry for not having been clear! What I meant is that the current implementation indeed never returns non-zero, and that's intended.

In previous iterations, there was argument handling logic in `homebrew-tap` that could return 1. However, as the function `homebrew-tap` is meant for zero-argument usage only, I removed that logic in e9b158cd5ca881cc3edd2c80f9cc537b11fcb1c4 and made sure at the call site that the function is only called when there are no `brew tap` arguments. I suppose your question is with the comment `falls back to cmd/tap.rb on a non-zero return` above, which I forgot to remove, so I removed it in ea011ae528bd1a67e3be56043b96c3da92073c75.

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:177]: > made sure at the call site that the function is only called when there are no `brew tap` arguments

Gotcha, I see that now. Might be a little clearer if this is above the wildcard commands and doesn't do `&& exit 0` but instead has `exit 0` be unconditional as there's no `return` status to check here.

Comment (ZhongRuoyu) [reply Library/Homebrew/brew.sh]: Done in c17154d2acebcfa085bd0df57d9e1902684c42c3.

**Review by MikeMcQuaid — APPROVED** [review]
Wonderful, thanks @ZhongRuoyu!

**Pull request merged**